programs/utilities/flower.sc

```diff
@@ -1,140 +1,299 @@
-//
-// Places the flower pattern that bonemeal creates when in flower forests.
-// large areas may cause quite a bit of lag to run.
-// Only places grass blocks + flowers in flower forest biomes.
-//
-// command usage:
-// /flower <position> print - prints what flower will generate at given position
-// /flower <position> create <radius> - places grass block with flowers on top in given radius (square area)
-// /flower <position> create <radius> only <flower> - same as create, but only places the chosen flower type
-//
-
-// flowers that generate in flower forest biome (in order)
-global_FLOWERS = ['dandelion', 'poppy', 'allium', 'azure_bluet', 'red_tulip', 'orange_tulip', 'white_tulip', 'pink_tulip', 'oxeye_daisy', 'cornflower', 'lily_of_the_valley'];
+// Utility for the flower pattern in flower forests
+// May cause quite a bit of lag if run on a large area
+
+// use legacy noise for versions before 1.18
+global_use_legacy_noise = system_info('game_major_target') < 18;
+
+// list of flowers that generate in flower forest biome (in order)
+global_flower_list = [
+  'dandelion',
+  'poppy',
+  'allium',
+  'azure_bluet',
+  'red_tulip',
+  'orange_tulip',
+  'white_tulip',
+  'pink_tulip',
+  'oxeye_daisy',
+  'cornflower',
+  'lily_of_the_valley'
+];
+global_flower_count = length(global_flower_list);
+
+// glass blocks that have a similar color to each of the flowers
+global_glass_list = [
+  'yellow_stained_glass',     // dandelion
+  'brown_stained_glass',      // poppy
+  'magenta_stained_glass',    // allium
+  'light_gray_stained_glass', // azure_bluet
+  'red_stained_glass',        // red_tulip
+  'orange_stained_glass',     // orange_tulip
+  'glass',                    // white_tulip
+  'pink_stained_glass',       // pink_tulip
+  'gray_stained_glass',       // oxeye_daisy
+  'blue_stained_glass',       // cornflower
+  'white_stained_glass'       // lily_of_the_valley
+];
+
+// noise constants
+// "seed" and "scale" values are from vanilla_worldgen/worldgen/configured_feature/flower_flower_forest.json
+// scale must be converted to a float then back to double
+// if using a custom noise, "firstOctave" must be 0 and "amplitudes" must be [1.0], otherwise this script doesn't support it
+global_flower_seed = 2345;
+global_scale_1 = 0.02083333395421505;
+global_scale_2 = global_scale_1 * 1.0181268882175227;
+global_double_perlin_amplitude = 0.8333333333333333;
+
+// these seeds are the first 2 nextLong calls of new Random(global_flower_seed) (for 2345: -1223197305642693068, -8087649459364435462)
+// then xor with "octave_0".hashCode() (1261148513)
+// todo: have these values automatically calculated from global_flower_seed
+global_perlin_seed_1 = -1223197304453310635;
+global_perlin_seed_2 = -8087649458443489125;
+
+// settings
+global_setting_ignore_biome = false;
+global_setting_glass_mode = false;
+global_setting_place_air_above = true;
```

Comment:
Since these are constant values, you can name them `global_SCALE_1`, `global_SCALE_2` and so on. 
This will separate them from other global values

Reply:
I checked a bunch of other scripts and none of them used uppercases for constants so I figured I'd just keep these named similarly to those

Reply:
ye I would agree with manyrandomthings here, best to keep code format the same all across